[WIP] Fix blocked container shutdown caused by non-reading attached client

asterite3 opened the pull request:
Fixes #41827

**- What I did**

Prevented container stop process from hanging indefinitely (causing all subsequent operations on that container to hang too) when there is a client attached to that container that is not reading container's output 

**- How I did it**

By adding a method `CleanContext` to `broadcaster.Unbuffered` that allows to put a timeout on cleanup (communicated using a `context`) - when that timeout expires, `Unbuffered` is forcibly cleaned by calling `Close` on its writers without waiting for their
`Write` operation to finish.

In more detail: we would like `CleanContext` to be able to call `Close` on a `WriteCloser` currently being blocked inside `Write` (in order to unblock it), while `Unbuffered.Write` is running. With current implementation of `Unbuffered` with a single lock it is unclear how this could be done: `Write` may hold the lock forever and there is no way to ask it to release. We also can't just ignore the lock and access `.writers` without holding it: `Unbuffered.Write` both uses and modifies `.writers`.

So the idea is to use _two_ locks instead, one of which must be held while updating `.writers` - and `Unbuffered.Write` will release it while blocking on its writers. Holding both locks means the same as holding `.mu` in original implementation - "full control" over this `Unbuffered` instance - and `.Add()` will do just that, as we don't need "forceful" semantics for it.

Also, we can use the fact that `.Clean()` does not resize `.writers` - it just sets it to `nil`. So, we can make `.writers` a pointer-to-slice instead of a slice - and have `Write` make a local copy of that pointer for itself - this will make sure `writers` in `Unbuffered.Write` remain valid. If `.CleanContext` has thrown old writers away, `Write` can change it in any way during eviction - it will just update an old (but valid) slice, that won't be used any more and will be picked up by GC after `Write` returns.

I am not sure whether current version of the fix is totally right, hence the WIP status.
I thought that closing writer that is stuck on write is better than just leaving it connected, giving up on some goroutine blocked on a `Write` to it. And it seems more aligned with what is done with `dio` in b5f28865efebb14c66d5580dfa7bf0634b5e3241.

Regarding how code is organized, maybe a simpler/more readable option would be to add a second (private) struct that will hold `writers` and a lock protecting their update [like this](https://gist.github.com/asterite3/95f5f7adabce0e60526de946c953f158). But, as you see, `Unbuffered.Write` explicitly takes/releases the lock of that private struct, so this sill kinda not perfect.

Another issue is whether it is correct to call `.Close()` while `.Write()` is running. As far as I understand, `io.WriteCloser` is not thread-safe by default, it may be an error to call it's methods simultaneously (and it does not guarantee to provide the desired effect: that `Close` will unblock `Write`). After reading the code it seems to me that most of the time writers in `Unbuffered` are `pkg/ioutil/BytesPipe`s, which are also thread-safe and work OK with interrupting `Write` with `Close`. But it is still possible to add non-thread-safe writer to `Unbuffered`, it's interface allows it and the way `container/stream.Config` uses it also allow to add any `WriteCloser` to it - on can call `Config`'s `Stdout()`, get `Unbuffered` object and call `Add` on it. This is different from how `cio.DirectIO` is made - if I understood [its code](https://github.com/containerd/containerd/tree/master/cio) correctly, its `closers` are not arbitrary `io.Closers`, but are always objects returned by `openFifo`.
So, maybe a more type-safe alternative would be to not use `Unbuffered` in `stream.Config` at all, but to add a new version of it, `broadcaster.BytesPipe`, that will ensure its writers _are_ `BytesPipe`s, and use it: [like this](https://github.com/moby/moby/compare/master...asterite3:add-broadcaster-bytespipe).

**- How to verify it**

Follow reproduction steps from #41827 and verify that container is no longer shown as running in `docker ps` after it stopped (and operations on it do not hang). 

**- Description for the changelog**

Fixed container stop sequence on daemon side being blocked indefinitely by an attached client that is not reading container's output.

**- A picture of a cute animal (not mandatory but encouraged)**
![s](https://user-images.githubusercontent.com/5569241/102715227-f4162200-42e4-11eb-8e96-dfa09fc69cd9.jpg)

asterite3 committed:
Add method CleanContext to broadcaster.Unbuffered

Method .Clean() of broadcaster.Unbuffered may block idefinitely
waiting for it's blocked writes to complete. Add new method
.CleanContext() that allows to forcefully close and remove writers
when context is done.
This is hard to implement with a single global lock for all
operations: there is no way to limit the time for which Write
will hold it and it is impossible for CleanContext() to use
.writers without holding a lock because .Write() may both read and
change .writers.
That's why:
 - introduce two locks instead of one. One of them
   (writersUpdateMu) should be held when updating writers slice -
   and .Write will release it while calling Write in writers. The
   second one (writersWriteMu) will prevent several Writes and
   Adds from happening simultaneously and from interfering with
   each other.
 - make writers a pointer-to-slice instead of slice so that
   CleanContext can assign writers to nil without interfering with
   Write (which will use it's own copy of that pointer).

Signed-off-by: Daniil Sigalov <asterite@seclab.cs.msu.ru>

asterite3 committed:
container: cleanup stdout/stderr with context

Add context argument to CloseStreams method of
container/stream.Config and methods Reset and CloseStreams
of container.Container: use that context to limit the time waiting
for stdout and stderr of stream.Config to cleanup (i. e. for all
their attached writers to close).

Signed-off-by: Daniil Sigalov <asterite@seclab.cs.msu.ru>

thaJeztah commented:
@cpuguy83 @tonistiigi ptal

payall4u commented:
We meet the same problem.
The stack of docker:

payall4u commented:
ping @cpuguy83 @tonistiigi  @fuweid

fuweid commented:
@payall4u would you send the whole goroutine dump log? thanks

payall4u commented:
> @payall4u would you send the whole goroutine dump log? thanks

[goroutine-stacks-2021-04-16T201709+0800.log](https://github.com/moby/moby/files/6348214/goroutine-stacks-2021-04-16T201709%2B0800.log)

asterite3 commented:
> I think it doesn't work. There are still some terminating pod. And docker goroutine dump here. @fuweid @asterite3
> 
> goroutine 6902181 [semacquire, 55 minutes]:
> sync.runtime_SemacquireMutex(0xc421fce7ac, 0xc421095f00)
> /usr/local/go/src/runtime/sema.go:71 +0x3f
> ...
> github.com/docker/docker/pkg/broadcaster.(*Unbuffered).CleanContext.func1(0xc424e08300, 0xc421fce7a0, 0xc4224b2e90)
> /go/src/github.com/docker/docker/pkg/broadcaster/unbuffered.go:117 +0x66
> created by github.com/docker/docker/pkg/broadcaster.(*Unbuffered).CleanContext
> /go/src/github.com/docker/docker/pkg/broadcaster/unbuffered.go:115 +0x9d

This stack belongs to goroutine that tries to do "regular" cleanup, with holding `writersWriteMu`, which may be held indefinitely by blocked `Write` call. Note that there is another path in `CleanContext` that does (forceful) cleanup without that lock if context expires.

@payall4u Can you please provide stacks for all other goroutines?
Is there a way to reproduce this situation when container shutdown hangs despite the fix in this PR?

payall4u commented:
IWe should use context with timeout when call `Unbuffered.Clean`, maybe.

> Can you please provide stacks for all other goroutines?

[goroutine-stacks-2021-06-02T165215+0800.log](https://github.com/moby/moby/files/6583447/goroutine-stacks-2021-06-02T165215%2B0800.log)


> Is there a way to reproduce this situation when container shutdown hangs despite the fix in this PR?

I'll try.

payall4u commented:
Maybe make the `unbuffered` one-time is a better way to fix the problem. It means never use again if cleaned.

asterite3 commented:
@payall4u after looking at stacktraces you provided I see that there is indeed a blocked call to `CleanContext` on line 126, it is blocked on `select` waiting for either normal cleanup to proceed, or for context to expire. But the call stack does not come from `handleContainerExit`, it comes from the call `execConfig.CloseStreams()` which happens inside the branch `if execConfig := c.ExecCommands.Get(ei.ProcessID); execConfig != nil {` of `ProcessEvent`. The respective stack trace is:
```
goroutine 6902165 [select, 55 minutes]:
github.com/docker/docker/pkg/broadcaster.(*Unbuffered).CleanContext(0xc421fce7a0, 0x25e03c0, 0xc42003c028, 0x513efb, 0x25e03c0)
    /go/src/github.com/docker/docker/pkg/broadcaster/unbuffered.go:126 +0x149
github.com/docker/docker/container/stream.(*Config).CloseStreams(0xc4264a2eb0, 0x25e03c0, 0xc42003c028, 0x25bc7a0, 0xc4200fc250)
    /go/src/github.com/docker/docker/container/stream/streams.go:103 +0x87
github.com/docker/docker/daemon/exec.(*Config).CloseStreams(0xc4252cb6c0, 0x25e0400, 0xc4214fb260)
    /go/src/github.com/docker/docker/daemon/exec/exec.go:84 +0x49
github.com/docker/docker/daemon.(*Daemon).ProcessEvent(0xc4203f0fc0, 0xc4219ea780, 0x40, 0x19e2799, 0x4, 0xc4219ea780, 0x40, 0xc4219ea7c0, 0x40, 0x890000294b, ...)
    /go/src/github.com/docker/docker/daemon/monitor.go:148 +0x37b
github.com/docker/docker/libcontainerd.(*client).processEvent.func1()
    /go/src/github.com/docker/docker/libcontainerd/client_daemon.go:694 +0x130
github.com/docker/docker/libcontainerd.(*queue).append.func1(0xc4219ea701, 0xc4217c9c20, 0xc421183a80, 0xc4233fd3e0, 0xc420411e88, 0xc4219ea780, 0x40)
    /go/src/github.com/docker/docker/libcontainerd/queue.go:26 +0x3a
created by github.com/docker/docker/libcontainerd.(*queue).append
    /go/src/github.com/docker/docker/libcontainerd/queue.go:22 +0x196
```

 So I think this is a slightly different, albeit similar, issue: caused by blocked output of **docker exec** (not `docker attach`). In my branch `exec.Config.CloseStreams` indeed calls `StreamConfig.CloseStreams()` with `context.Background()` - which means there is no timeout. https://github.com/moby/moby/blob/111fbfd0489948825ed1435929976f4d92a2b975/daemon/exec/exec.go#L84
I passed `context.Background` to all calls to `stream.Config.CloseStreams` except one in `Container.Reset` to (somewhat conservatively) maintain the behaviour before the patch. Probably other calls are worth adding a timeout too.

I was able to reproduce the situation like one in provided stacktraces with following steps:
1. `docker run --name testing --rm -it ubuntu bash`. This time container entrypoint does not log a lot, it's just bash.
2. (in another console) `docker exec testing yes | /bin/sleep infinity`. This creates `docker exec` that is blocked on writing its output
3. (in yet another console) `docker kill testing`. This command will hang for some time, then exit with status 0. Docker daemon will write to log `Container 9b7b2000f0af failed to exit within 10 seconds of kill - trying direct SIGKILL` at that point.
4. After that, container will be still shown as running in `docker ps`. In process list (`ps aux`) it can be seen that main process of the container is not running any more, but it's containerd shim is still running. And situation in goroutine stack dumps of dockerd will like the one provided by @payall4u (there will be a `CleanContext` blocked on `select`)

asterite3 committed:
daemon: close exec config streams with timeout

Closing streams of exec config can be blocked idefinitely when
writes of output streams are blocked - similarly to what may
happen to container's own output streams when there is a
non-reading attached client. This may block container shutdown,
preventing it from completing. Use context with timeout to limit
the time waiting for execConfig.CloseStreams to complete during
container shutdown.

Signed-off-by: Daniil Sigalov <asterite@seclab.cs.msu.ru>

jl2501 commented:
any updates here? (asking for a friend) ;)

laurazard commented:
Hiya @asterite3! I've just bumped into this PR while looking at https://github.com/moby/moby/issues/46078, and found your PR. I'd like to get it merged! I've rebased your PR, but I have some comments. Do have availability to address some review comments? Otherwise I'd be happy to take over and try to get it merged :)

laurazard commented on daemon/start.go:
```diff
@@ -130,7 +130,7 @@ func (daemon *Daemon) containerStart(ctx context.Context, daemonCfg *configStore
 				log.G(ctx).Errorf("%s: failed saving state on start failure: %v", container.ID, err)
 			}
-			container.Reset(false)
+			container.Reset(context.Background(), false)
```
why not pass the context here?

asterite3 replied:
I did not pass it only because I tried to be somewhat careful when doing changes in this PR, trying to change how the daemon works only where necessary (at least until explicitly discussed with maintainers)

asterite3 replied:
I have updated the code: now `ctx` is passed here. Also, context is now passed to `CloseStreams` in `daemon/exec.go` [here](https://github.com/asterite3/moby/blob/13f9fd2/daemon/exec.go#L190)

laurazard commented on container/container.go:
```diff
@@ -803,7 +803,7 @@ type rio struct {
 	i.IO.Close()
 
-	return i.sc.CloseStreams()
+	return i.sc.CloseStreams(context.Background())
```
We should probably have a default timeout here – otherwise, cases like https://github.com/moby/moby/issues/46078#issuecomment-1667840537

where the blocking call to `CloseStreams` comes from
```
goroutine 1174391 [sync.Mutex.Lock, 508 minutes]:
sync.runtime_SemacquireMutex(0x5557d0292f56?, 0xa0?, 0xa3?)
	/usr/local/go/src/runtime/sema.go:77 +0x26
sync.(*Mutex).lockSlow(0xc00a2a1800)
	/usr/local/go/src/sync/mutex.go:171 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:90
github.com/docker/docker/pkg/broadcaster.(*Unbuffered).Clean(0xc00a2a1800)
	/go/src/github.com/docker/docker/pkg/broadcaster/unbuffered.go:42 +0x3a
github.com/docker/docker/container/stream.(*Config).CloseStreams(0xc009e59a40)
	/go/src/github.com/docker/docker/container/stream/streams.go:102 +0x105
github.com/docker/docker/container.(*rio).Close(0xc00900a558)
	/go/src/github.com/docker/docker/container/container.go:806 +0x37
github.com/docker/docker/vendor/github.com/containerd/containerd.(*task).Delete(0xc00107b540, {0x5557d11f8148, 0xc0097176e0}, {0x0, 0x0, 0x5557d265df40?})
	/go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/task.go:328 +0x2ab
```

are still going to block.

asterite3 replied:
You're right. However, my testing shows that it is even more complicated than that: during my testing `libcontainerd/remote.(*task).Delete()` calls `t.io.Wait()` [here](https://github.com/moby/moby/blob/4d67489/vendor/github.com/containerd/containerd/task.go#L339)
this is a call to [rio.Wait()](https://github.com/moby/moby/blob/4d674897f3748a28347f2233237dbfae7c18cdc1/container/container.go#L809)
this, in turn, calls `i.sc.Wait(context.Background())` - which hangs, probably because containerd-shim does not exit, probably because it waits for the receiver to read the data it sends (container's output that we do not read because attached reader is not reading). Maybe `rio.Wait()` should also accept a context and pass that to `i.sc.Wait()`

asterite3 replied:
After some more thinking, I see the following ways to change `handleContainerExit` to fix this blocking:
* move the deletion after waiting and closing IO, [like this](https://gist.github.com/asterite3/691d439187e11c6bfac4edf82e2c2f3e). This is the simplest fix and it seems to be in the spirit of containerd commit [fc535efe5845aa99f98b7fac58526b1b06c102ab](https://github.com/containerd/containerd/commit/fc535efe5845aa99f98b7fac58526b1b06c102ab) (see my other comment below for more details about that commit). This will give contanerd shim around 2 seconds to transfer the remaining container output to docker daemon.
* A more conservative approach: call `tsk.Delete(ctx)` *before* closing the IO, but, instead of waiting indefinitely for that call to return, only wait for some time, and close IO streams then ( `tsk.Delete(ctx)`  can be called in a separate goroutine for that). This second option makes the code of `handleContainerExit` larger and more complicated, but in this case the code will behave closer to how it was before the fix
 
According to my testing, both ways fix the problem
I can't say I fully understand the implications of the first option because I'm not super familiar with Docker architecture, whether it will work OK in all cases. I myself did not find other consequences beside the descreased timeout for IO completion (timeouts of contexts passed to `StreamConfig.Wait(ctx)` and `c.Reset(ctx, false)` can, of course, be increased to mitigate that). It would be great to get someone else's opinion here

asterite3 replied:
I have added a commit that implements the second, more conservative, approach. It uses a timeout of 3 seconds while waiting for `tsk.Delete(ctx)` before proceeding to closure of IO streams. In addition with a 2-second timeout passed to `StreamConfig.Wait`, dockerd will give around 5 seconds for container output to be transferred from containerd to readers. If we set this timeout to 4 or more seconds, `docker kill` will start complaining that container did not exit in 10 seconds and that it sends a "direct SIGKILL". Maybe that's OK but is the reason I chose a 3-second time and not more.

It can be noted that there is a context with 30-second timeout passed to `tsk.Delete(ctx)` and we could wait for that to expire, but my experiments showed that, if that context is cancelled, `tsk.Delete(ctx)` fails and containerd shim is not deleted, it remains running without a child process (seemingly forever).

asterite3 replied:
A different approach to fix this is to address the issue on containerd side. It is evident that, although `tsk.Delete` accepts a context, its cancellation is not fully respected and [`t.io.Wait()`](https://github.com/containerd/containerd/blob/2aeac9a/task.go#L337) may block forever. I'm not sure how the fix on containerd should be implemented though.

Interface `cio.IO` does not support passing contexts to its method calls, so we would have to either update that interface or add a method `WaitContext` to `rio` that's not in that interface and test for that in `Delete` using a type assertion.
Also, if I'm not mistaken, if `ctx` passed to `Delete` is cancelled, an RPC call `t.client.TaskService().Delete` would fail. So maybe `Delete` should pass another context, with lesser timeout, to that waiting method.
(And obviously this would require a PR to containerd and synchronizing the vendored version)

asterite3 commented:
Hi @laurazard!

Wow, it's been a long time since this PR was created -)

> I have some comments. Do have availability to address some review comments?

Yes, I do have some time and I would like to try to address your comments to get this merged!

laurazard commented on daemon/monitor.go:
```diff
@@ -189,9 +191,11 @@ func (daemon *Daemon) ProcessEvent(id string, e libcontainerdtypes.EventType, ei
 
-			if err := execConfig.CloseStreams(); err != nil {
+			ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+			if err := execConfig.CloseStreams(ctx); err != nil {
```
this change doesn't fix the issue afaict because the blocking call to `CloseStreams` doesn't come from here, but earlier from `handleContainerExit` which calls `tsk.Delete(ctx)` which calls `io.Close()`, which doesn't take a context/will still block forever.

asterite3 replied:
Yes, you're right. The code has changed since this PR was created and now it hangs earlier

asterite3 replied:
Btw, I did a bit more analysis on why the code was not getting stuck on that `Delete` earlier, but is getting stuck there now.

In past this `handleContainerExit` was calling [`Process.Delete`](https://github.com/containerd/containerd/blob/157fe6a/process.go#L204), not [`Task.Delete`](https://github.com/containerd/containerd/blob/2aeac9a/task.go#L299)

(to be more exact, call chain was [`handleContainerExit`](https://github.com/moby/moby/blob/b5f863c67e6ffbaadedb94c885c3c50c625e4eb8/daemon/monitor.go#L30) -> [`client.DeleteTask`](https://github.com/moby/moby/blob/b5f863c67e6ffbaadedb94c885c3c50c625e4eb8/libcontainerd/remote/client.go#L465) -> `process.Delete`)

As far as I understand, the difference is that `Process.Delete` calls `p.client.TaskService().DeleteProcess()` *before* waiting for and closing `io` ([it happens here](https://github.com/containerd/containerd/blob/157fe6a/process.go#L225)), while `Task.Delete` will call `t.client.TaskService().Delete()` only *after* it does waiting and closing io ([here](https://github.com/containerd/containerd/blob/2aeac9a/task.go#L339)). My guess it that, after the deletion containerd shim exits and channels leading to dockerd are closed, so `io.Wait()` won't get stuck

I also noticed that `Process.Delete` was not always working like that, initially it also was calling `p.io.Wait()` *before* calling `DeleteProcess`. But this was changed by containerd commit [`fc535efe5845aa99f98b7fac58526b1b06c102ab`](https://github.com/containerd/containerd/commit/fc535efe5845aa99f98b7fac58526b1b06c102ab). It belongs to containerd/containerd#1394 - it seems that `Process.Delete` working that way was causing some tests to sometimes deadlock

laurazard commented:
PTAL @thaJeztah @vvoland @corhere. I started investigating this in https://github.com/moby/moby/issues/46078#issuecomment-1667840537 before I found related issues/this PR, so there's some more context about this issue there.

corhere commented on pkg/broadcaster/unbuffered.go:
```diff
@@ -1,49 +1,141 @@
+// call Close while Write may be in progress in order to forcefully close
+// writers
+func (w *Unbuffered) CleanContext(ctx context.Context) error {
+	// there is essentially a race between normal cleaning (done while holding
+	// w.writersWriteMu) and forceful one - cleaningUnderway is used to tell
+	// who won the race and avoid double cleanup
```
The implementation could be greatly simplified, eliminating the race, [if `w.writersWriteMu` was a 1-buffered `chan struct{}` instead of a classical mutex.](https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view)
```go
select {
case <-w.writersWriteMu:
	defer func() { w.writersWriteMu <- struct{}{} }()
case <-ctx.Done():
}
w.cleanWithWriteLock()
return ctx.Err()
```

asterite3 replied:
Thanks for the idea! -)

I'll look into this

asterite3 replied:
@corhere I have tried this idea. `CleanContext` indeed looks much better after this change. However, there is an issue here - unlike a mutex, a channel requires initialization: it should be created with `make()`. `broadcaster.Unbuffered` does not currently have a constructor, so we have 2 options here:
* create `w.writersWriteMu` channel on demand. This will create a new race condition - a race for the `w.writersWriteMu` field itself. For example, if two goroutines call `.Add()` on a fresh `Unbuffered` instance at the same time, there will be a race in assignment of the created `w.writersWriteMu`. So, we will need an additional lock that protects this field. An implementation with this additional lock can look like this https://gist.github.com/asterite3/eb938f97f58a02a14916cebfaa174373 (this is not well tested yet, but I think it would more-or-less like this)
* add a constructor function `NewUnbuffered` to `broadcaster` package. Then, all users of this package should be updated to start using this function. I am not entirely sure how hard is that. On the one hand, in moby repository it seems that there is only one place where `Unbuffered` is created: `container/stream/streams.go`. On the other hand, this is a public package and other users may exist (although by looking at [godoc](https://pkg.go.dev/github.com/docker/docker/pkg/broadcaster?tab=importedby) it seems only forks of moby are using it). Are there any guarantees about the stability of packages in `pkg/`? If adding a (mandatory) constructor function is OK, it seems a better option.

@laurazard what do you think ?

asterite3 replied:
You're right, `sync.Once` can be used instead of an additional lock (`writersWriteMuMu` in my implementation). And on-demand initialization will look much simpler then.

asterite3 replied:
Is this an implementation variant we want to proceed with?
I have switched to it in this PR.

(The previuos implementation that does not require an extra `sync.Once` but where `CleanContext` is complicated and uses atomics is [left here](https://gist.github.com/asterite3/069b11ee1d41ecbe9e926593c1aa30b1) just in case)

asterite3 replied:
Btw, `sync.Once` can also be used to simplify `CleanContext` in variant where `writersWriteMu` is a regular `sync.Mutex`: it can [look like this](https://gist.github.com/asterite3/44a3c6ab93b43888e0ba8c2142ce5d02#file-unbuffered-go-L105). Also, as far as I understand, after this change `CleanContext` is not just simpler, the semantics are changed so that `CleanContext` never returns until `w.cleanWithWriteLock` is done - which is a right thing I guess. And in my initial implementation `CleanContext` could return before `w.cleanWithWriteLock` returned (i. e. before cleanup is done), which seems to be a bug.

laurazard replied:
Sorry, have been out on leave for a while will take a look. Generally, re:
> Is this an implementation variant we want to proceed with?

I think the answer is yes, since it seems both cleaner and more correct. I'll try to review it soon :)

asterite3 replied:
> Sorry, have been out on leave for a while will take a look. Generally, re:
> 
> > Is this an implementation variant we want to proceed with?
> 
> I think the answer is yes, since it seems both cleaner and more correct. I'll try to review it soon :)

Okay! -)

The variant with atomics is certainly worse and should not be chosen, but I encourage you to also check out the variant linked from my previous comment, which also uses `sync.Once`, but inside `CleanContext`, and where `writersWriteMu` is a `sync.Mutex`

asterite3 commented on pkg/broadcaster/unbuffered.go:
```diff
@@ -1,49 +1,141 @@
+	// there is essentially a race between normal cleaning (done while holding
+	// w.writersWriteMu) and forceful one - cleaningUnderway is used to tell
+	// who won the race and avoid double cleanup
+	var cleaningUnderway int32 = 0
```
Sure thing! -)

`atomic.Bool` was not yet added to Go language when I was making this PR. Now that it's added, I will switch to it, it is indeed better

asterite3 committed:
daemon: fix tsk.Delete blocking during container exit

A call to tsk.Delete() in handleContainerExit can block
idefinitely due to blocked output streams of the container. To fix
that, close container output streams if Delete call does not
return for more than 3 seconds (and continue waiting for Delete to
complete after that).

Signed-off-by: Daniil Sigalov <asterite@seclab.cs.msu.ru>

LanderV-Vintecc commented:
Any update on this? I suspect this would fix an issue we have.

laurazard commented:
Right, it's been a while and I forgot about this until recently. I'll take some time to look at it today and see if @asterite3 is still around/active to carry it over.

mar0us commented:
Is there any news on this issue? I'm facing the same problem and I don't have any ideas how to solve it.
https://github.com/moby/moby/issues/46078

@laurazard
